--- wpa_supplicant-0.7.1/src/drivers/driver_wext.c.old	2011-08-16 10:24:40.311098001 -0400
+++ wpa_supplicant-0.7.1/src/drivers/driver_wext.c	2011-08-16 10:25:01.555098001 -0400
@@ -1731,19 +1731,6 @@
 			}
 			return;
 		}
-		/*
-		 * Clear the BSSID selection and set a random SSID to make sure
-		 * the driver will not be trying to associate with something
-		 * even if it does not understand SIOCSIWMLME commands (or
-		 * tries to associate automatically after deauth/disassoc).
-		 */
-		for (i = 0; i < 32; i++)
-			ssid[i] = rand() & 0xFF;
-		if (wpa_driver_wext_set_bssid(drv, null_bssid) < 0 ||
-		    wpa_driver_wext_set_ssid(drv, ssid, 32) < 0) {
-			wpa_printf(MSG_DEBUG, "WEXT: Failed to set bogus "
-				   "BSSID/SSID to disconnect");
-		}
 	}
 }
 
